Spigot-Server-Patches/0594-Damage-with-any-cause.patch

```diff
@@ -0,0 +1,437 @@
++                        return DamageSource.arrow((EntityArrow) attackingEntity, projectile.getShooter());
++                    } else if (entityType == EntityTypes.SNOWBALL || entityType == EntityTypes.EGG ||
++                        entityType == EntityTypes.ENDER_PEARL || entityType == EntityTypes.POTION) {
++                        return DamageSource.projectile(attackingEntity, projectile.getShooter());
++                    } else if (entityType == EntityTypes.FIREWORK_ROCKET) {
++                        return DamageSource.a((EntityFireworks)attackingEntity, projectile.getShooter());
++                    } else if (entityType == EntityTypes.WITHER_SKULL) {
++                        return DamageSource.a((EntityWitherSkull)attackingEntity, projectile.getShooter());
++                    }
++                    return DamageSource.projectile(projectile, projectile.getShooter());
++                }
++                return DamageSource.GENERIC;
++            case SUFFOCATION:
++                return DamageSource.STUCK;
++            case FALL:
++                return DamageSource.FALL;
++            case FIRE:
++                return DamageSource.FIRE;
++            case FIRE_TICK:
++                return DamageSource.BURN;
++            case MELTING:
++                return CraftEventFactory.MELTING;
++            case LAVA:
++                return DamageSource.LAVA;
++            case DROWNING:
++                return DamageSource.DROWN;
++            case VOID:
++                return DamageSource.OUT_OF_WORLD;
++            case LIGHTNING:
++                return DamageSource.LIGHTNING;
++            case STARVATION:
++                return DamageSource.STARVE;
++            case POISON:
++                return CraftEventFactory.POISON;
++            case MAGIC:
++                return DamageSource.MAGIC;
++            case WITHER:
++                return DamageSource.WITHER;
++            case FALLING_BLOCK:
++                return DamageSource.FALLING_BLOCK;
++            case THORNS:
++                if (attackingEntity == null) {
++                    return DamageSource.GENERIC;
++                }
++                return DamageSource.a(attackingEntity);
++            case DRAGON_BREATH:
++                return DamageSource.DRAGON_BREATH;
++            case CUSTOM:
++                return DamageSource.GENERIC;
++            case FLY_INTO_WALL:
++                return DamageSource.FLY_INTO_WALL;
++            case HOT_FLOOR:
++                return DamageSource.HOT_FLOOR;
++            case CRAMMING:
++                return DamageSource.CRAMMING;
++            case DRYOUT:
++                return DamageSource.DRYOUT;
++        }
++        return DamageSource.GENERIC;
++    }
++
++    public static EntityDamageEvent.DamageCause damageSourceToDamageCause(DamageSource source, @Nullable Entity damager, Entity damagee) {
++        if (source == DamageSource.ANVIL || source == DamageSource.FALLING_BLOCK) {
```

Comment:
This stuff should be on the NMS `DamageSource`; that'd easily cause a patch conflict if it's changed.